cpp/src/arrow/acero/asof_join_node.cc

```diff
@@ -1520,6 +1520,9 @@ class AsofJoinNode : public ExecNode {
       // Plan has already aborted.  Do not start process thread
       return Status::OK();
     }
+    #ifndef ARROW_ENABLE_THREADING
+      return Status::ExecutionError("ASOF join requires threading enabled");
```

Comment:
Probably move this at the beginning of this method? Also use a `#else` clause around the regular implementation.

And I would probably make this `Status::NotImplemented`.